### DISCUSSION TITLE
Loading large tree sequences with pyslim

### DISCUSSION THREAD

**Discussion by santaci** [tskit-dev/pyslim#303] [Q&A]
Perhaps this is a naive question given the nature of ts files, but I was wondering if there was a way to load tree sequence files in a multi-threaded or parallel manner using pyslim? 

Unfortunately I have  tree sequence files that I need to load several times if I want to work on them in parallel. Granted, they are about 5-6 Gb each, but the loading alone takes anywhere from 10-15 min. Is there a more clever way to load these tree sequence files? I don't have issues with ts files that are lighter (e.g. ~1 Gb) but these 5-6 Gb seem to take way longer to load.

**Comment by bhaller** [comment]
Nowadays `tskit.load()` is preferred over `pyslim.load()`, as I understand it, and @petrelharp will of course know better but I think `pyslim.load()` might just be a pass-through to tskit.  So this sounds like an issue that might be better filed against tskit?  It sounds like your big tree sequence file might be a good test case for them to optimize for; it would be very useful if you could upload it somewhere so that it can be downloaded for testing and profiling.  Of course it's a very large file, so it might be a bit difficult to find a place to host it; but if you don't have an obvious place to put it, your IT department should be able to lend a hand.  I doubt the tskit folks will want to parallelize their load algorithm any time soon (that sort of work is tremendously complex, and a load time of 15 minutes is a bit painful but not the end of the world), but there may certainly be other ways to speed up the load that would be worth looking at.  Tagging @benjeffery and @jeromekelleher.

**Reply by jeromekelleher** [reply]
This sounds waaaaaay too long! How long does it take to load one of these files with ``tskit.load()`` @santaci ?

**Reply by benjeffery** [reply]
The easiest way to check would be to run `time tskit info filename.trees` on the command line.

**Reply by santaci** [reply]
Hi guys, thanks a bunch for answering my question so promptly!

So turns out that one of the reasons why it was taking about 10 minutes to load was because I was loading two ts files consecutively (one with neutral mutations and one without).
But my mistake aside, loading the 6.3 Gb trees (the one with neutral mutations), took with `pyslim.load()`: `0:03:33.674013`.
That same file performing `time tskit info add_grad_4_mu.trees` took `2m28.050s`. 

I apologise for adding on the other 7 mins which are from loading another ts file + loading python modules I imagine. But overall it looks like `tskit.load()` might be a bit faster than `pyslim.load` as pointed out by Ben Haller.

I'm also happy to provide my tree sequence should it still be of use/interest for benchmarking—I do have somewhere I can host it, should it prove useful :-)

Again, thanks for pointing out that that `tskit.load()` is the way to go these day!

**Reply by santaci** [reply]
Ah but now I also remember that the reason I avoided using `tskit.load() ` was because I was losing slim-related attributes...

**Reply by bhaller** [reply]
> Ah but now I also remember that the reason I avoided using `tskit.load() ` was because I was losing slim-related attributes...

If you're using the current tskit and pyslim it ought to work fine, I _think_.  Again, @petrelharp will know better than I; possibly I have misunderstood.  It used to be that using `pyslim.load()` was necessary to get the SLiM metadata properly, but I believe that is no longer the case.  Anyway, you might try it.

**Reply by jeromekelleher** [reply]
Can you give us the full output of ``time tskit info add_grad_4_mu.trees`` please? I wonder how much of the time is user vs system time.

**Reply by santaci** [reply]
Sure, it was as follows:
`real	2m28.050s`
`user	1m3.057s`
`sys	0m21.504s`

I'll also try updating my pyslim and tskit in a different environment and give it a go as well.

**Reply by jeromekelleher** [reply]
OK, so a lot of the time is waiting for IO, which makes sense for such large files. Still, 1 minute of CPU time seems excessive for loading. I wonder where the time is being spent.

**Reply by petrelharp** [reply]
Wait up!   `pyslim.load( )` no longer exists (I think it produces a deprecation error), and certainly did do some slow stuff that was made obsolete by tskit. So - @santaci, could you update your pyslim and tskit, then try again? Have a look at [this page](https://tskit.dev/pyslim/docs/stable/previous_versions.html) for insructions, but there's very little to change.

**Reply by benjeffery** [reply]
@santaci If you could share the file that would be good. In the mean time can you paste in the output of `tskit info`? Seeing the sizes of tables would be interesting. I assume most of the time is building the various indexes that are created on load, but it still seems excessive.

**Reply by santaci** [reply]
@benjeffery I've posted here the output of tskit info using version `0.3.4`:
`sequence_length:  68937975.0`
`trees:            16447189`
`samples:          20016158`
`individuals:      10015389`
`nodes:            30489023`
`edges:            130099466`
`sites:            22599769`
`mutations:        22599769`
`migrations:       0`
`populations:      4`
`provenances:      5`

I've also updated my tskit in a separate environment to version `0.5.2` and obtain the following tables with a slower run time:

<img width="324" alt="Screenshot 2022-09-29 at 12 58 44" src="https://user-images.githubusercontent.com/29976164/193014307-0b5719ae-9a67-4d96-8e56-25c8e5dcd73a.png">


`real	2m5,385s`
`user	1m41,021s`
`sys	0m24,209s`

I can share the tree sequence via our IT storage hosting here: https://sid.erda.dk/share_redirect/aNOb2oRrJm
Just be sure to rename it as *.trees file when downloading.
